lib/ansible/modules/cloud/docker/docker_volume.py

```diff
@@ -0,0 +1,245 @@
+
+
+class TaskParameters(DockerBaseClass):
+    def __init__(self, client):
+        super(TaskParameters, self).__init__()
+        self.client = client
+
+        self.volume_name = None
+        self.driver = None
+        self.driver_options = None
+        self.labels = None
+        self.force = None
+        self.debug = None
+
+        for key, value in iteritems(client.module.params):
+            setattr(self, key, value)
+
+
+class DockerVolumeManager(object):
+
+    def __init__(self, client):
+        self.client = client
+        self.parameters = TaskParameters(client)
+        self.check_mode = self.client.check_mode
+        self.results = {
+            u'changed': False,
+            u'actions': []
+        }
+        self.diff = self.client.module._diff
+
+        self.existing_volume = self.get_existing_volume()
+
+        state = self.parameters.state
+        if state == 'present':
+            self.present()
+        elif state == 'absent':
+            self.absent()
+
+    def get_existing_volume(self):
+        volumes = self.client.volumes()
+        for volume in volumes[u'Volumes']:
+            if volume['Name'] == self.parameters.volume_name:
+                return volume
+
+        return None
+
+    def has_different_config(self):
+        """
+        Return the list of differences between the current parameters and the existing volume.
+
+        :return: list of options that differ
+        """
+        differences = []
+        if self.parameters.driver and self.parameters.driver != self.existing_volume['Driver']:
+            differences.append('driver')
+        if self.parameters.driver_options:
+            if not self.existing_volume.get('Options'):
+                differences.append('driver_options')
+            else:
+                for key, value in iteritems(self.parameters.driver_options):
+                    if (not self.existing_volume['Options'].get(key) or
+                            value != self.existing_volume['Options'][key]):
+                        differences.append('driver_options.%s' % key)
+        if self.parameters.labels:
+            existing_labels = self.existing_volume.get('Labels', {})
+            all_labels = set(self.parameters.labels) | set(existing_labels)
+            for label in all_labels:
+                if existing_labels.get(label) != self.parameters.labels.get(label):
+                    differences.append('labels.%s' % label)
+
+        return differences
+
+    def create_volume(self):
+        if not self.existing_volume:
+            if not self.check_mode:
+                resp = self.client.create_volume(self.parameters.volume_name,
+                                                 driver=self.parameters.driver,
+                                                 driver_opts=self.parameters.driver_options,
+                                                 labels=self.parameters.labels)
```

Comment:
Same as above.

Wrap calls to self.client.X in try/except block. On APIError call self.client.fail() with an error message and exception details.